packages/embark/src/lib/core/events.js

```diff
@@ -45,22 +54,64 @@ EventEmitter.prototype.request = function() {
 
   log("requesting: ", requestName);
   warnIfLegacy(requestName);
-  return this.emit('request:' + requestName, ...other_args);
+  const listenerName = 'request:' + requestName;
+
+  // if we don't have a command handler set for this event yet,
+  // store it and fire it once a command handler is set
+  if (!this.listeners(listenerName).length) {
+    if(!toFire[listenerName]) {
+      toFire[listenerName] = [];
+    }
+    toFire[listenerName].push(other_args);
+    return;
+  }
+
+  return this.emit(listenerName, ...other_args);
 };
 
 EventEmitter.prototype.setCommandHandler = function(requestName, cb) {
   log("setting command handler for: " + requestName);
   let listener = function(_cb) {
     cb.call(this, ...arguments);
   };
+  const listenerName = 'request:' + requestName;
+
   // unlike events, commands can only have 1 handler
-  this.removeAllListeners('request:' + requestName);
-  return this.on('request:' + requestName, listener);
+  _removeAllListeners.call(this, listenerName);
+
+  // if this event was requested prior to the command handler
+  // being set up, call the callback. Do not remove this listener
+  // so any 
+  const prematureListenerArgs = cloneDeep(toFire[listenerName]);
+  if (prematureListenerArgs) {
+    delete toFire[listenerName];
+    this.on(listenerName, listener);
```

Comment:
You can remove this line and the return below and only register the `on` after. I don't think it will cause an issue

Reply:
To support a use case where an event is requested inside of a listener callback, it has to be bound **before** the initial request is called. An example of this (from the tests):
```
      let testData = { isTest: true, manipulatedCount: 0 };
      events.request(testEventName, testData, (dataFirst) => {
        expect(dataFirst.isTest).to.be.true;
        expect(dataFirst.manipulatedCount).to.equal(1);
        expect(dataFirst.isAnotherTest).to.be.undefined;
      });
      events.request(testEventName, testData, (dataSecond) => {
        expect(dataSecond.isTest).to.be.true;
        expect(dataSecond.manipulatedCount).to.equal(2);
        expect(dataSecond.isAnotherTest).to.be.undefined;

        dataSecond.isAnotherTest = true;
        events.request(testEventName, dataSecond, (dataThird) => {
          expect(dataThird.isTest).to.be.true;
          expect(dataThird.manipulatedCount).to.equal(3);
          expect(dataThird.isAnotherTest).to.be.true;
          done();
        });
      });
      events.setCommandHandler(testEventName, (options, cb) => {
        expect(options.isTest).to.be.true;
        options.manipulatedCount++;
        cb(options);
      });
```
If this line was removed, the listener would be attached **after** the initial callback was called, rendering the nested callback unanswerable.